src/main/java/com/github/sardine/impl/SardineImpl.java

```diff
@@ -1028,12 +1012,33 @@ public boolean exists(String url) throws IOException
 	protected HttpResponse execute(HttpRequestBase request)
 			throws IOException
 	{
+		return execute(context, request, null);
+	}
+
+	/**
+	 * Common method as single entry point responsible fo request execution
+	 * @param context clientContext to be used when executing request
+	 * @param request Request to execute
+	 * @param maybeResponseHandler can be null if you need raw HttpResponse or not null response handler for result handling.
+	 * @param <T> will return raw HttpResponse when maybeResponseHandler is null or value reslved using provided ResponseHandler instance
+	 * @return value resolved using response handler or raw HttpResponse when maybeResponseHandler is null
+	 */
+	protected <T> T execute(HttpClientContext context, HttpRequestBase request, ResponseHandler<T> maybeResponseHandler)
+			throws IOException
+	{
 		try
 		{
 			// Clear circular redirect cache
-			this.context.removeAttribute(HttpClientContext.REDIRECT_LOCATIONS);
-			// Execute with no response handler
-			return this.client.execute(request, this.context);
+			context.removeAttribute(HttpClientContext.REDIRECT_LOCATIONS);
+
+			if (maybeResponseHandler != null)
```

Comment:
Not sure if I understand what you mean.

From inside SardineImpl you are using both versions of httpClient execute method - with response handler and without it. Execute method with response handler will close InputStream to remote resource and you will not be able to read from it.
Take a look at com.github.sardine.impl.SardineImpl#get(java.lang.String, java.util.List<org.apache.http.Header>) there is a reason why you are not passing VoidResponseHandler directly into httpClient execute method. If you'd do that then you will not be able to read file content from ContentLengthInputStream because underlying InputStream will be closed by org.apache.http.util.EntityUtils#consume called from inside of httpClient.execute method.

If I use response handler (EmptyResponseHandler or whatever) in all calls then end users will not be able to download (GET) files because HttpClient will close InputStream and you will not be able to read from it.
Loading data into byte array is out of question because you will not be able to handle huge files correctly.
Storing data somewhere in /tmp is not good solution as well because you might want to process files on the fly without saving file locally first.